src/apps/service/apps/jobs/submissions/async.clj

```diff
@@ -0,0 +1,118 @@
+               :error "All HT Analysis Path Lists must have the same number of paths."}))
+    path-lists))
+
+(defn- get-path-list-contents
+  [user path]
+  (try+
+    (when (seq path) (data-info/get-path-list-contents user path))
+    (catch Object _
+      (log/error (:throwable &throw-context)
+                 "job submission failed: Could not get file contents of path list input.")
+      (throw+))))
+
+(defn- get-path-list-contents-map
+  [user paths]
+  (into {} (map (juxt identity (partial get-path-list-contents user)) paths)))
+
+(defn- map-slice
+  [m n]
+  (->> (map (fn [[k v]] (vector k (nth v n))) m)
+       (into {})
+       (remove-nil-values)))
+
+(defn- map-slices
+  [m]
+  (let [max-count (apply max (map (comp count val) m))]
+    (mapv (partial map-slice m) (range max-count))))
+
+
+(defn- substitute-param-values
+  [path-map config]
+  (->> (map (fn [[k v]] (vector k (get path-map v v))) config)
+       (into {})))
+
+(defn- format-submission-in-batch
+  [submission job-number path-map]
+  (let [job-suffix (str "analysis-" (inc job-number))]
+    (assoc (update-in submission [:config] (partial substitute-param-values path-map))
+      :group      (config/jex-batch-group-name)
+      :name       (str (:name submission) "-" job-suffix)
+      :output_dir (ft/path-join (:output_dir submission) job-suffix))))
+
+(defn- submit-job-in-batch
+  [apps-client user submission job-number path-map]
+  (try+
+    (submit/submit-and-register-private-job apps-client
+                                            user
+                                            (format-submission-in-batch submission job-number path-map))
+    (catch Object _
+      (log/error (:throwable &throw-context)
+                 "batch job submission failed.")
+      {:status jp/failed-status})))
+
+(defn- preprocess-batch-submission
+  [submission output-dir parent-id]
+  (assoc submission
+    :output_dir           output-dir
+    :parent_id            parent-id
+    :create_output_subdir false))
+
+(defn submit-batch-jobs
+  "Asynchronously submits sub-jobs for a parent batch job,
+   where `ht-paths` is a list of paths of the HT Path List input files,
+   `submission` is the original parent submission map,
+   `output-dir` is the path to the output folder created for the parent job,
+   and `parent-id` is the parent job's ID to use for each sub-job's `parent_id` field.
+   A notification will be sent to the user on success or on failure to submit all sub-jobs."
+  [apps-client user ht-paths submission output-dir parent-id]
+  (future
+    (try+
+      (let [path-lists    (validate-path-lists (get-path-list-contents-map user ht-paths))
+            path-maps     (map-slices path-lists)
+            submission    (preprocess-batch-submission submission output-dir parent-id)
+            job-stats     (->> path-maps
+                               (map-indexed (partial submit-job-in-batch apps-client user submission))
```

Comment:
I started working on `CORE-9014` and I realized that `submit-job-in-batch` will have to check the parent job's status before submitting each sub-job, in case the user tries to cancel the parent in the middle of batch submissions.

I'll go ahead and make that change for this PR.